File: 239-sliding-window-maximum/sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        int n=nums.size();
        vector<int> ans;
        set<pair<int,int>,greater<pair<int,int>>> st;
        int max_ele=0;
        for(int i=0;i<k;i++){
            st.insert({nums[i],i});
        }
        int i=0,j=k;
        while(j < n){
            auto [ele,ind]=*st.begin();
            ans.push_back(ele);
            st.erase({nums[i],i});
            st.insert({nums[j],j});
            i++;
            j++;
        }
        auto [ele,ind]=*st.begin();
        ans.push_back(ele);
        return ans;
    }
};

File: 239-sliding-window-maximum/sliding-window-maximum.py
class Solution:
    def maxSlidingWindow(self, nums: list[int], k: int) -> list[int]:
        q=deque()
        ans=[]
        ind=0
        or_k=k
        while k:
            if ind==0:
                q.append(ind)
            else:
                while q and nums[q[-1]] <= nums[ind]:
                    q.pop()
                q.append(ind)
            ind+=1
            k-=1
        ans.append(nums[q[0]])
        for i in range(or_k,len(nums)):
            while q and  q[0] <= (i-or_k):
                q.popleft()
            while q and nums[q[-1]] <= nums[i]:
                q.pop()
            q.append(i)
            ans.append(nums[q[0]])
        return ans
        